src/google/cloud/ndb/global_cache.py

```diff
@@ -160,3 +165,110 @@ def compare_and_swap(self, items, expires=None):
             current_value = self.cache.get(key)
             if watch_value == current_value:
                 self.cache[key] = (new_value, expires)
+
+
+_Pipeline = collections.namedtuple("_Pipeline", ("pipe", "id"))
+
+
+class RedisCache(GlobalCache):
+    """Redis implementation of the :class:`GlobalCache`.
+
+    This is a synchronous implementation. The idea is that calls to Redis
+    should be fast enough not to warrant the added complexity of an
+    asynchronous implementation.
+
+    Args:
+        redis (redis.Redis): Instance of Redis client to use.
+    """
+
+    @classmethod
+    def from_environment(cls):
+        """Generate a class:`RedisCache` from an environment variable.
+
+        This class method looks for the ``REDIS_CACHE_URL`` environment
+        variable and, if it is set, passes its value to ``Redis.from_url`` to
+        construct a ``Redis`` instance which is then used to instantiate a
+        ``RedisCache`` instance.
+
+        Returns:
+            Optional[RedisCache]: A :class:`RedisCache` instance or
+                :data:`None`, if ``REDIS_CACHE_URL`` is not set in the
+                environment.
+        """
+        url = os.environ.get("REDIS_CACHE_URL")
+        if url:
+            return cls(redis_module.Redis.from_url(url))
+
+    def __init__(self, redis):
+        self.redis = redis
+        self.pipes = {}
+
+    def get(self, keys):
+        """Implements :meth:`GlobalCache.get`."""
+        res = self.redis.mget(keys)
+        return res
+
+    def set(self, items, expires=None):
+        """Implements :meth:`GlobalCache.set`."""
+        self.redis.mset(items)
+        if expires:
+            for key in items.keys():
+                self.redis.expire(key, expires)
+
+    def delete(self, keys):
+        """Implements :meth:`GlobalCache.delete`."""
+        self.redis.delete(*keys)
+
+    def watch(self, keys):
```

Comment:
Is watch() part of the original ndb's spec for caching? I don't understand how they made this work with the memcached api.

Reply:
``watch`` and ``compare_and_swap`` are equivalent to using ``get`` and ``set`` in memcached using CAS tokens. Had to generalize the concept a bit to work with Redis. I also did away with the CAS acronym because it's a bit opaque. (I had to do some Googling when I first encountered it.)